### PR TITLE
Add task to calculate ensemble mean and variance from background state

### LINKED ISSUE #365: Adding ensemble mean
From a discussion with @rtodling, I will be testing and adding an ensemble mean to the workflow.

### PR DESCRIPTION
**Pull request by metdyn**
## Description

A step towards LETKF is to calculate ensemble mean and variance:
- https://github.com/GEOS-ESM/swell/issues/365

Alternatively, this functionality may be used as a simple, independent tool.

Tested on SCU17:
```
swell create localensembleda  -s slurmfile.yaml -p nccs_discover_sles15

slurm_directives_tasks:
  RunJediLocalEnsembleDaExecutable:
    geos_atmosphere:
      nodes: 1
      ntasks-per-node: 36
  GetEnsembleMeanVariance:
    all:
      nodes: 1
      ntasks-per-node: 36
```

## Dependencies

None

## Impact

It is part of the localensembleda suite.

### CODE REVIEW

**Comment by rtodling** [comment]
This seems to work - it runs fine, but I am confused as to what's preventing the localensembleda to go beyond simply  the ensemean background ... after all, this is behaving as if it were the ensemblemean suite but it's not!

**Comment by Dooruk** [comment]
@rtodling and @metdyn this is missing the last commit from the `develop` branch.

Please see following changes. We are abandoning the static `StageJedi` task to be able to use relative paths and instead running `StageJediCycle` at each cycle, which puts static files in the `{cycle_dir}`: 

https://github.com/GEOS-ESM/swell/compare/feature/submit/ens_mean...develop

@metdyn, please notice the changes in `flow.cylc` for `3dvar_atmos` and `3dfgat_atmos`, they need to be replicated for the `localensembleda` suite.

**Comment by metdyn** [comment]
> @rtodling and @metdyn this is missing the last commit from the `develop` branch.
> 
> Please see following changes. We are abandoning the static `StageJedi` task to be able to use relative paths and instead running `StageJediCycle` at each cycle, which puts static files in the `{cycle_dir}`:
> 
> [feature/submit/ens_mean...develop](https://github.com/GEOS-ESM/swell/compare/feature/submit/ens_mean...develop)
> 
> @metdyn, please notice the changes in `flow.cylc` for `3dvar_atmos` and `3dfgat_atmos`, they need to be replicated for the `localensembleda` suite.

Thanks, @Dooruk, I will take a look and make appropriate changes to the PR to be update with other suites and relative path.

**Comment by metdyn** [comment]
Hi @Dooruk @asewnath,

I know the ctest `swell test  code_tests` fails at 
https://github.com/GEOS-ESM/swell/blob/develop/src/swell/utilities/scripts/task_question_dicts.py#L134-L135

Error message is
```
FAIL: test_dictionary_comparison (swell.test.code_tests.question_dictionary_comparison_test.QuestionDictionaryTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/gpfsm/dnb07/projects/p234/yyu11/br_jedi_scu17/install_swell/src/swell/swell_ens_mean/src/swell/test/code_tests/question_dictionary_comparison_test.py", line 26, in test_dictionary_comparison
    assert tq_dicts_rc == 0
AssertionError
```

So I printed out the string of the dict into files:
```
yaml.dump(question_dict_str_in)
yaml.dump(dict_to_write_str)
```

diff diff in.txt   out.txt

Then I print out the first two lines of differences of the two files:

```
7,25c7,24
<   \ are the analysis variables?\n  tasks:\n  - GenerateBClimatology\n  - PrepareAnalysis\n\
<   \  - RunJediConvertStateSoca2ciceExecutable\n  - RunJediVariationalExecutable\n\
  \  - GetEnsembleMeanVariance\n  type: string-check-list\n\nbackground_error_model:\n\

---
>   \ are the analysis variables?\n  tasks:\n  - GenerateBClimatology\n  - GetEnsembleMeanVariance\n\
>   \  - PrepareAnalysis\n  - RunJediConvertStateSoca2ciceExecutable\n  - RunJediVariationalExecutable\n\
  \  type: string-check-list\n\nbackground_error_model:\n  ask_question: false\n \
```

The order between `GetEnsembleMeanVariance` and `PrepareAnalysis` is swap. 
So the dictionary comparison using the print out string becomes a problem when new tasks is added into task dir,
which changed glob.glob file list.
There should be a better way for the task questions yaml comparison.

Screenshot 2024-08-02 at 11 49 19 AM


Meanwhile I tried to print out the difference between the two dictionaries:
> set1 = set(dict1.items())
> set2 = set(dict2.items())
> set1 ^ set2
(https://stackoverflow.com/questions/32815640/how-to-get-the-difference-between-two-dictionaries-in-python)
There comes an error that the dictionary is not hashable.  I did not go further.

**Comment by metdyn** [comment]
I think this error means I need to regenerate `tasks/task_questions.yaml`, because new tasks are added in. 
In this logic, then the test_dictionary_comparison function would be a good place as a generator.  Or I would manually check tasks/task_questions.yaml.  How was this done before?

**Comment by Dooruk** [comment]
@metdyn, let me look into this branch and see what is going on. I will suggest an update to this branch afterwards as a merge. It's a simple unit test issue but I also notice some potential `Stage` task problems. 

If you change this following line in the source code (setting `$LOG_INFO` to 1 in the environment didn't work for me) and build it again, tests will provide more details to show you a temporary YAML file with the proper `task_questions.yaml`.
https://github.com/GEOS-ESM/swell/blob/465990f83f167f6820a4994d1af407804823d0f2/src/swell/test/code_tests/code_tests.py#L28-L29

Sorry, this is all very obscure without documentation but this is also how I learned these things..

**Comment by metdyn** [comment]
@Dooruk thanks and that is the key! I was about to ask you why the logger.info did not work in this test code file, but logger.test() worked in another file.

**Comment by Dooruk** [comment]
@rtodling, can you confirm `RunJediEnsembleMeanVariance` works on your end so we can merge this?

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py [hunk 1]
```diff
@@ -27,6 +27,7 @@ def execute(self):
         # Jedi application name
         # ---------------------
         jedi_application = 'localensembleda'
+        jedi_ensmeanvariance_application = 'ensmeanvariance'
```
Comment (Dooruk) [inline src/swell/tasks/run_jedi_local_ensemble_da_executable.py:30]: Is this still necessary after `RunJediEnsembleMean` task is implemented?

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py [hunk 2]
```diff
@@ -37,6 +38,8 @@ def execute(self):
         observations = self.config.observations()
         jedi_forecast_model = self.config.jedi_forecast_model(None)
         generate_yaml_and_exit = self.config.generate_yaml_and_exit(False)
+        ensmean_only = self.config.ensmean_only()
+        ensmeanvariance_only = self.config.ensmeanvariance_only()
```
Comment (Dooruk) [inline src/swell/tasks/run_jedi_local_ensemble_da_executable.py:42]: Same comment regarding `RunJediEnsembleMean`

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py [hunk 3]
```diff
@@ -115,6 +118,10 @@ def execute(self):
                                     self.config.local_ensemble_save_posterior_mean_increment())
         self.jedi_rendering.add_key('local_ensemble_save_posterior_ensemble_increments',
                                     self.config.local_ensemble_save_posterior_ensemble_increments())
+        self.jedi_rendering.add_key('ensmean_only',
+                                    self.config.ensmean_only())
+        self.jedi_rendering.add_key('ensmeanvariance_only',
+                                    self.config.ensmeanvariance_only())
```
Comment (Dooruk) [inline src/swell/tasks/run_jedi_local_ensemble_da_executable.py:124]: same ^ `RunJediEnsembleMean`

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py [hunk 4]
```diff
@@ -180,16 +187,29 @@ def execute(self):
...
+                self.logger.info('Running ' + jedi_ensmeanvariance_executable_path +
+                                 ' with '+str(np)+' processors.')
+                self.logger.info('Running ensmean_only')
+                run_executable(self.logger, self.cycle_dir(), np,
+                               jedi_ensmeanvariance_executable_path,
+                               jedi_config_file, output_log_file)
+            else:
+                self.logger.info('Running '+jedi_executable_path+' with '+str(np)+' processors.')
+                run_executable(self.logger, self.cycle_dir(), np, jedi_executable_path,
+                               jedi_config_file, output_log_file)
```
Comment (Dooruk) [inline src/swell/tasks/run_jedi_local_ensemble_da_executable.py:212]: Same comment regarding `RunJediEnsembleMean` ^

File: src/swell/tasks/task_questions.yaml [hunk 5]
```diff
@@ -198,10 +199,37 @@ ensemble_num_members:
...
+  models:
+  - geos_atmosphere
+  options:
+  - true
+  - false
+  prompt: Calculate ensemble mean and variance only?
+  tasks:
+  - RunJediLocalEnsembleDaExecutable
+  type: boolean
+
```
Comment (Dooruk) [inline src/swell/tasks/task_questions.yaml:232]: same ^

File: src/swell/test/suite_tests/localensembleda-tier1.yaml [hunk 6]
```diff
@@ -3,23 +3,27 @@ jedi_build_method: use_existing
...
+    npx_proc: 2
+    npy_proc: 2
     cycle_times:
     - T00
-    - T06
-    - T12
-    - T18
+#    - T06
+#    - T12
+#    - T18
```
Comment (Dooruk) [inline src/swell/test/suite_tests/localensembleda-tier1.yaml]: You could keep the `cycle_times` uncommented but keep a single `start_cycle_time` and `end_cycle_time` so T06, T12, and T18 won't be triggered.

File: src/swell/utilities/render_jedi_interface_files.py [hunk 7]
```diff
@@ -84,6 +84,8 @@ def __init__(self, logger, experiment_root, experiment_id, cycle_dir, cycle_time
             'local_ensemble_save_posterior_mean',
             'local_ensemble_save_posterior_mean_increment',
             'local_ensemble_solver',
+            'ensmean_only',
+            'ensmeanvariance_only',
```
Comment (Dooruk) [inline src/swell/utilities/render_jedi_interface_files.py:88]: same^

File: .gitignore [hunk 8]
```diff
@@ -2,6 +2,9 @@
 *.swp
 *DS_Store*
 ._*
+*~
+.#*
```
Comment (Dooruk) [inline .gitignore:6]: Why were these changed?

`._*` are Mac temporary files.

**Review by Dooruk — CHANGES_REQUESTED** [review]
I added a few comments. There is some duplicate functionality because `RunJediEnsembleMean` already handles the options where you put inside `RunJediLocalDaExecutable`. Simply taking them out should suffice.

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py (hunk 1 above)
Comment (metdyn) [reply src/swell/tasks/run_jedi_local_ensemble_da_executable.py:30]: The `RunJediEnsembleMean` is an external test.  This statement is a separate one that can be used after LETKF run.

File: .gitignore (hunk 8 above)
Comment (metdyn) [reply .gitignore:6]: These are modified file copies left by Emacs

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py (hunk 2 above)
Comment (metdyn) [reply src/swell/tasks/run_jedi_local_ensemble_da_executable.py:42]: same as above, to compute mean and variance at the end of this code.

File: src/swell/tasks/task_questions.yaml (hunk 5 above)
Comment (metdyn) [reply src/swell/tasks/task_questions.yaml:232]: If we donot need it later on, we will remove it.

File: src/swell/test/suite_tests/localensembleda-tier1.yaml (hunk 6 above)
Comment (metdyn) [reply src/swell/test/suite_tests/localensembleda-tier1.yaml]: Thanks, that will also work!  I have changed the T06,12,18 back.

**Review by metdyn — COMMENTED** [review]
I created another PR to help with enHX with packets. 
https://github.com/GEOS-ESM/swell/pull/405

File: src/swell/tasks/run_jedi_local_ensemble_da_executable.py (hunk 1 above)
Comment (Dooruk) [reply src/swell/tasks/run_jedi_local_ensemble_da_executable.py:30]: Ok thanks, this is now clear to me after today's meeting.

**Review by Dooruk — APPROVED** [review]
This is good to go on my end, let's wait on @rtodling .

**Review by rtodling — APPROVED** [review]
This works for me - it fails in the HofX, but I understand this is another issue to be tackled elsewhere.

**Pull request merged**